Add RPC internal helper that overrides the default pickler

zzzwen opened the pull request:
Summary: To enable share_memory over RPC, add an internal helper that overrides the default RPC pickler

Test Plan: ```python test/test_rpc_spawn RpcTestWithSpawn.test_use_rpc_pickler```

Differential Revision: D18598974

mrshenli commented on torch/distributed/rpc/api.py:
```diff
@@ -8,13 +8,27 @@
 
+_default_pickler = _internal_rpc_pickler
+
+@contextlib.contextmanager
```
Shall we add a test for this?

mrshenli commented on torch/distributed/rpc/api.py:
```diff
@@ -8,13 +8,27 @@
+@contextlib.contextmanager
+def _UseRPCPickler(rpc_pickler):
+    r"""
+    rpc_pickler: (.internal._InternalRPCPickler) Overrides the default RPC pickler
```
You mean `_internal_rpc_pickler`?

zzzwen replied:
.internal._InternalRPCPickler is the type

mrshenli commented on torch/distributed/rpc/api.py:
```diff
@@ -8,13 +8,27 @@
 from .internal import _internal_rpc_pickler, PythonUDF
 
 import datetime
+import contextlib
```
move before datetime

mrshenli commented on test/rpc_test.py:
```diff
@@ -14,6 +14,7 @@
 from common_utils import load_tests
 from dist_utils import INIT_METHOD_TEMPLATE, TEST_CONFIG, dist_init
 from torch.distributed.rpc.internal import PythonUDF, _internal_rpc_pickler
+from torch.distributed.rpc.api import _UseRPCPickler
```
move one line above

mrshenli commented on torch/distributed/rpc/api.py:
```diff
@@ -8,13 +8,27 @@
+_default_pickler = _internal_rpc_pickler
+
+@contextlib.contextmanager
+def _UseRPCPickler(rpc_pickler):
```
BTW, any reason for using camel here?

mrshenli reviewed (commented):
LGTM. Added some nits. Will stamp when test pass.

rohan-varma commented:
Could you update the test plan with how an OSS user could run this test? Something like `python test/test_rpc_spawn RpcTestWithSpawn. test_use_rpc_pickler` would be great

Commit:
Add RPC internal helper that overrides the default pickler (#30107)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/30107

To enable share_memory over RPC, add an internal helper that overrides the default RPC pickler

Test Plan: `buck test mode/dev-nosan //caffe2/test:rpc_spawn -- test_use_rpc_pickler`

Differential Revision: D18598974

fbshipit-source-id: d5a6583e94f1c98fa0db2c1d6f0e66ff5c6141d9

mrshenli reviewed (approved):
Tests passed but there are conflicts in `rpc_test.py`. Please address and make sure tests pass after that before landing.

pietern reviewed (changes requested):
Removing approval until the conflict is resolved and tests pass.

zzzwen commented:
Github does not pickup the rebased version. Close this PR and re-exporting...

zzzwen commented:
Reopen as https://github.com/pytorch/pytorch/pull/30173
